mitiq/zne/inference.py

```diff
@@ -202,6 +237,262 @@ def get_expectation_values(self) -> np.ndarray:
+            np.ndarray,
+        )
+
+    def run(
+        self,
+        qp: QPROGRAM,
+        executor: Union[Callable[..., float], Callable[..., List[float]]],
+        scale_noise: Callable[[QPROGRAM, float], QPROGRAM],
+        num_to_average: int = 1,
+    ) -> "BatchedFactory":
+        """Computes the expectation values at each scale factor and stores them
+        in the factory. If the executor returns a single expectation value, the
+        circuits are run sequentially. If the executor is batched and returns
+        a list of expectation values (one for each circuit), then the circuits
+        are sent to the backend as a single job. To detect if an executor is
+        batched, it must be annotated with a return type that is one of the
+        following:
+
+            * Iterable[float]
+            * List[float]
+            * Sequence[float]
+            * Tuple[float]
+            * numpy.ndarray
+
+        Args:
+            qp: Quantum circuit to run.
+            executor: A "single executor" (1) or a "batched executor" (2).
+                (1) A function which inputs a single circuit and outputs a
+                single expectation value of interest.
+                (2) A function which inputs a list of circuits and outputs a
+                list of expectation values (one for each circuit). A batched
+                executor can also take an optional "kwargs_list" argument to
+                set a list of keyword arguments (one for each circuit). This
+                is necessary only if the factory is initialized using the
+                optional "shot_list" parameter.
+
+            scale_noise: Noise scaling function.
+            num_to_average: The number of circuits executed for each noise
+                scale factor. This parameter can be used to increase the
+                precision of the "executor" or to average the effect of a
+                non-deterministic "scale_noise" function.
+        """
+        self.reset()
+        self._batch_populate_instack()
+
+        # Get all noise-scaled circuits to run
+        to_run = self._generate_circuits(qp, scale_noise, num_to_average)
+
+        # Get the list of keywords associated to each circuit in "to_run"
+        kwargs_list = self._get_keyword_args(num_to_average)
+
+        if self._is_executor_batched(executor):
+            if all([kwargs == {} for kwargs in kwargs_list]):
+                res = executor(to_run)
+            else:
+                res = executor(to_run, kwargs_list=kwargs_list)
+        else:
+            res = [
+                executor(circ, **kwargs)  # type: ignore
+                for circ, kwargs in zip(to_run, kwargs_list)
+            ]
+
+        # Reshape "res" to have "num_to_average" columns
+        res = np.array(res).reshape((-1, num_to_average))
+
+        # Average the "num_to_average" columns
+        self._outstack = np.average(res, axis=1)
+
+        return self
+
+    def run_classical(
```

Comment:
renaming this from `iterate` is a backwards-incompatible change. I would recommend adding back an `iterate` method to `Factory` that (1) raises a `DeprecationWarning` saying that `iterate` will be removed in `v0.4.0`, and (2) then calls `run_classical`

Reply:
Deprecated iterate.